File: 1-100/13.roman-to-integer.cpp
class Solution {
public:
    int romanToInt(const string& s) {
        int ret = 0;
        int preNum = 0;
        int curNum = 0;
        for (const auto& c : s)
        {
            curNum = GetRomanCharInt(c);
            if (preNum < curNum)
            {
                ret -= preNum;
            }
            else
            {
                ret += preNum;
            }
            preNum = curNum;
        }
        ret += preNum;

        return ret;
    }

    int GetRomanCharInt(const char c)
    {
        switch (c)
        {
        case 'I': return 1;
        case 'V': return 5;
        case 'X': return 10;
        case 'L': return 50;
        case 'C': return 100;
        case 'D': return 500;
        case 'M': return 1000;
        default: return 0;
        }
    }
};

File: 1-100/13.roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        ret = 0
        numDict = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        preNum = 0
        curNum = 0
        for c in reversed(s):
            curNum = numDict.get(c, 0)
            if curNum < preNum:
                ret -= curNum
            else:
                ret += curNum
                preNum = curNum
        
        return ret
